Minimum Spanning Tree/MinimumSpanningTree.playground/Sources/Graph.swift

```diff
@@ -0,0 +1,55 @@
+
+// Undirected edge
+public struct Edge<T>: CustomStringConvertible {
+  public let vertex1: T
+  public let vertex2: T
+  public let weight: Int
+
+  public var description: String {
+    return "[\(vertex1)-\(vertex2), \(weight)]"
+  }
+}
+
+
+// Undirected weighted graph
+public struct Graph<T: Hashable>: CustomStringConvertible {
+
+  public private(set) var edgeList: [Edge<T>]
+  public private(set) var vertices: Set<T>
+  public private(set) var adjList: [T: [(vertex: T, weight: Int)]]
+
+  public init() {
+    edgeList = [Edge<T>]()
+    vertices = Set<T>()
+    adjList = [T: [(vertex: T, weight: Int)]]()
+  }
+
+  public var description: String {
+    var description = ""
+    for edge in edgeList {
+      description += edge.description + "\n"
+    }
+    return description
+  }
+
+  public mutating func addEdge(vertex1 v1: T, vertex2 v2: T, weight w: Int) {
+    edgeList.append(Edge(vertex1: v1, vertex2: v2, weight: w))
+    vertices.insert(v1)
+    vertices.insert(v2)
+    if adjList[v1] == nil {
+      adjList[v1] = [(vertex: v2, weight: w)]
+    } else {
+      adjList[v1]!.append((vertex: v2, weight: w))
+    }
+
+    if adjList[v2] == nil {
```

Comment:
@Xiangxin I was wondering the same, but after talking to some of my peers, it will not internally make a new copy unless adjList[v1] was mutated. You can go with either  nil coalescing or what you originally had with adjList[v1] == nil check. I prefer nil checks over if let _ = adjList[v1] since the intent is more clear. Thanks!